### PR TITLE
New type backend architecture

### LINKED ISSUE #14: tally very slow
Here is a case where tally is very slow. Interestingly, only a small change in the constraint generation/simplification part cause this digression.

Test case:

```
-module(test3).

-compile(export_all).

-spec foo2 (a) -> 1; (b) -> 2.
foo2(a) -> 1;
foo2(b) -> 2.
```

Branch sw/guard-fix shows the problem. The problematic constraints are as follows:

```
[{csubty, "var A1@0":6:1, a, $2},
 {csubty, "tuple constructor":6:1, {$2}, $0},
 {csubty,
  "t // {...}":7:1,
  ($0 /\ not({a} /\ {any()})) /\ ({b} /\ {any()}),
  {$9}},
 {csubty,
  "t // {...}":7:1,
  ($0 /\ not({a} /\ {any()})) /\ ({b} /\ {any()}),
  {$8}},
 {csubty,
  "exhaustiveness check":6:1,
  $0,
  {a} /\ {any()} | {b} /\ {any()}},
 {csubty,
  "t // {...}":7:1,
  ($0 /\ not({a} /\ {any()})) /\ ({b} /\ {any()}),
  {$7}},
 {csubty, "t // {...}":6:1, $0 /\ ({a} /\ {any()}), {$6}},
 {csubty, "t // {...}":6:1, $0 /\ ({a} /\ {any()}), {$5}},
 {csubty, "t // {...}":6:1, $0 /\ ({a} /\ {any()}), {$4}},
 {csubty,
  "t // {...}":7:1,
  ($0 /\ not({a} /\ {any()})) /\ ({b} /\ {any()}),
  {$10}},
 {csubty, "t // {...}":6:1, $0 /\ ({a} /\ {any()}), {$3}}]
```

When you do `git reset --hard ec9cbf94a97bc85523c2a049ddafa92c58b05cd7`, performance is back to normal.

### PR DESCRIPTION
**Pull request by albsch**
Implemented the subtyping and tally algorithms from scratch resulting a new and efficient architecture.

Fixes #14 and #8.

### CODE REVIEW

**Comment by skogsbaer** [comment]
@albsch I pushed one trivial commit into your PR.

File: test_files/tycheck_simple.erl [hunk 1]
```diff
@@ -404,11 +404,12 @@ tuple_05_fail(X) ->
 -spec use_atom(atom()) -> atom().
 use_atom(X) -> X.
 
+
 -spec inter_01(integer()) -> integer()
             ; (atom()) -> atom().
 inter_01(X) ->
     case X of
-        _ when is_integer(X) -> X + 1;
+        _ when is_integer(X) -> X;
```
Comment (skogsbaer) [inline test_files/tycheck_simple.erl:412]: why?

File: test_files/tycheck_simple.erl [hunk 2]
```diff
@@ -417,15 +418,15 @@ inter_01(X) ->
             ; (integer()) -> integer().
 inter_02(X) ->
     case X of
-        _ when is_integer(X) -> X + 1;
+        _ when is_integer(X) -> X;
```
Comment (skogsbaer) [inline test_files/tycheck_simple.erl:421]: why?

File: test/prop_opt_set.erl [hunk 3]
```diff
@@ -1,279 +0,0 @@
...
-    subtype(T, S).
-
-%%
-%%equiv2(S, T) ->
-%%  subtype_old(S, T) andalso
-%%    subtype_old(T, S).
-%%
-empty(S) ->
-  equiv(S, {predef, none}).
-
```
Comment (skogsbaer) [inline test/prop_opt_set.erl:277]: why did you delete these tests?

File: test/subty_tests.erl [hunk 4]
```diff
@@ -518,16 +514,7 @@ bug1_test() ->
...
-empty_tuple_test() ->
-  O2 = {intersection,
-    [
-      {tuple, [{singleton,b}]},
-      {tuple,[{singleton,a}]}
-    ]},
 
-  true = subty:is_empty(O2, none),
-
-  ok.
```
Comment (skogsbaer) [inline test/subty_tests.erl:530]: Some tests disappeared. Why?

File: src/typing_test.erl [hunk 5]
```diff
@@ -92,6 +92,22 @@ should_run(Name, {exclude,Set}) -> not sets:is_element(Name, Set).
 
 simple_test_() ->
   WhatNot = [
+    % FIXME soundness
```
Comment (skogsbaer) [inline src/typing_test.erl:95]: what's the problem here?

File: src/ast.erl [hunk 6]
```diff
@@ -467,40 +475,195 @@ is_predef_alias_name(N) ->
...
+                _ -> {union, Filtered}
+            end
     end.
 
 -spec mk_negation(ast:ty()) -> ast:ty().
 mk_negation({predef, any}) -> {predef, none};
 mk_negation({predef, none}) -> {predef, any};
 mk_negation(T) -> {negation, T}.
+
+erlang_ty_to_ast(X) ->
```
Comment (skogsbaer) [inline src/ast.erl:546]: Imo, this function should be moved somewhere else. For me, the ast module is a specification and should not contain much code. I already thought about moving the smart constructors elsewhere.

File: src/typing_test.erl [hunk 7]
```diff
@@ -13,9 +13,9 @@ check_ok_fun(Filename, Tab, Decl = {function, L, Name, Arity, _}, Ty) ->
     try
         typing:check(Ctx, Decl, Ty)
     catch
-        throw:{ety, ty_error, Msg} ->
-            io:format("~s: Type checking ~w/~w in ~s failed but should succeed: ~s",
-                      [ast:format_loc(L), Name, Arity, Filename, Msg]),
+        throw:{ety, ty_error, _Msg} ->
+%%            io:format("~s: Type checking ~w/~w in ~s failed but should succeed: ~s",
+%%                      [ast:format_loc(L), Name, Arity, Filename, Msg]),
```
Comment (skogsbaer) [inline src/typing_test.erl:18]: remove code if no longer needed.

File: src/typing_test.erl [hunk 8]
```diff
@@ -44,8 +44,8 @@ check_fail_fun(Filename, Tab, Decl = {function, L, Name, Arity, _}, Ty) ->
     Ctx = typing:new_ctx(Tab, error),
     try
         typing:check(Ctx, Decl, Ty),
-        io:format("~s: Type checking ~w/~w in ~s succeeded but should fail",
-                  [ast:format_loc(L), Name, Arity, Filename]),
+%%        io:format("~s: Type checking ~w/~w in ~s succeeded but should fail",
+%%                  [ast:format_loc(L), Name, Arity, Filename]),
```
Comment (skogsbaer) [inline src/typing_test.erl:48]: remove code if no longer needed.

File: test/prop_opt_set.erl (hunk 3 above)
Comment (albsch) [reply test/prop_opt_set.erl:277]: Property-based tests will be ported separately.

File: src/ast.erl (hunk 6 above)
Comment (albsch) [reply src/ast.erl:548]: Will be split up into `ast.erl`  and  `ast_lib.erl`

File: src/typing_test.erl (hunk 5 above)
Comment (albsch) [reply src/typing_test.erl:95]: Internal tally bugs that need to be fixed before merging.

File: test_files/tycheck_simple.erl (hunk 1 above)
Comment (albsch) [reply test_files/tycheck_simple.erl]: restored old tests

File: src/typing_test.erl (hunk 5 above)
Comment (albsch) [reply src/typing_test.erl]: Substitution was not applied correctly for the newly implemented lists.

**Review by skogsbaer — APPROVED** [review]
